machine/v1/types_nutanixprovider.go

```diff
@@ -154,6 +164,156 @@ type NutanixResourceIdentifier struct {
 	Name *string `json:"name,omitempty"`
 }
 
+// NutanixGPUIdentifierType is an enumeration of different resource identifier types for GPU entities.
+// +kubebuilder:validation:Enum:=Name;DeviceID
+type NutanixGPUIdentifierType string
+
+const (
+	// NutanixGPUIdentifierName identifies a GPU by Name.
+	NutanixGPUIdentifierName NutanixGPUIdentifierType = "Name"
+
+	// NutanixGPUIdentifierDeviceID identifies a GPU by device ID.
+	NutanixGPUIdentifierDeviceID NutanixGPUIdentifierType = "DeviceID"
+)
+
+// NutanixGPU holds the identity of a Nutanix GPU resource in the Prism Central
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'DeviceID' ?  has(self.deviceID) : !has(self.deviceID)",message="deviceID configuration is required when type is DeviceID, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Name' ?  has(self.name) : !has(self.name)",message="name configuration is required when type is Name, and forbidden otherwise"
+// +union
+type NutanixGPU struct {
+	// type is the identifier type of the GPU device.
+	// Valid values are Name and DeviceID.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Type NutanixGPUIdentifierType `json:"type"`
+
+	// deviceID is the GPU device ID with the integer value.
+	// +optional
+	// +unionMember
+	DeviceID *int32 `json:"deviceID,omitempty"`
+
+	// name is the GPU device name
+	// +optional
+	// +unionMember
+	Name *string `json:"name,omitempty"`
```

Comment:
Did we already confirm that there's no way to restrict the valid characters here?

Reply:
I think the GPU device name is also from the GPU vendor, I am not sure if it has any valid characters restriction. We used it here just to identify the GPU device.

Reply:
I was thinking this would be the name as it appears to the host. 

I'm a bit confused as to how this might work to be honest, what would happen if you had two identical GPUs from the same vendor, how would you identify which one was which when attaching? Is it possible to attach two identical GPUs?

Reply:
If you review the implementation PR https://github.com/openshift/machine-api-provider-nutanix/pull/77, you will see that we use the GPU name or deviceID to look for a GPU attached to one of the PE hosts and in "UNUSED" status. If found, it is used when creating the VM. The GPU name and deviceID can be the same, for example:

host (name: karbonmsp-gpu02-1, UUID: 52362f6f-fa6f-471b-9ad0-9811fbd213de)
"gpu_list": [{"status": "UNUSED", "vendor": "NVIDIA", "num_virtual_display_heads": 0, "assignable": true, "num_vgpus_allocated": 0, "pci_address": "0000:ca:00.0", "name": "Ampere 40", "frame_buffer_size_mib": 49152, "index": 0, "max_resolution": "None", "mode": "PASSTHROUGH_COMPUTE", "fraction": 0, "guest_driver_version": "None", "device_id": 8757, "max_instances_per_vm": 0}]

host (name: karbonmsp-gpu03-1, UUID: 4c38cd3f-6fc1-4cf8-af51-3410902c90aa)
"gpu_list": [{"status": "UNUSED", "vendor": "NVIDIA", "num_virtual_display_heads": 0, "assignable": true, "num_vgpus_allocated": 0, "pci_address": "0000:ca:00.0", "name": "Ampere 40", "frame_buffer_size_mib": 49152, "index": 0, "max_resolution": "None", "mode": "PASSTHROUGH_COMPUTE", "fraction": 0, "guest_driver_version": "None", "device_id": 8757, "max_instances_per_vm": 0}]